Update etcd-druid documentation and enhance kind-up script to start a local kind registry container (#889)
gardener/etcd-druid: docs/deployment/getting-started-locally/getting-started-locally.md

>
> ```bash
> > ./hack/kind-up.sh -h
>   usage: kind-up.sh [Options]
>   Options:
>     --cluster-name  <cluster-name>   Name of the kind cluster to create. Default value is 'etcd-druid-e2e'
>     --skip-registry                  Skip creating a local docker registry. Default value is false.
>     --feature-gates <feature-gates>  Comma separated list of feature gates to enable on the cluster.
> ```

## 02-Setting up etcd-druid

### Configuring etcd-druid

Prior to deploying `etcd-druid`, it can be configured via CLI-args and environment variables. 

* To configure CLI args you can modify [`charts/druid/values.yaml`](https://github.com/gardener/etcd-druid/blob/3383e0219a6c21c6ef1d5610db964cc3524807c8/charts/druid/values.yaml).  For e.g. if you wish to `auto-reconcile` any change done to `Etcd` CR then you should set `enableEtcdSpecAutoReconcile` to true. By default this will be switched off.
* `DRUID_E2E_TEST=true` : sets specific configuration for etcd-druid for optimal e2e test runs, like a lower sync period for the etcd controller.

### Deploying etcd-druid

> Any variant of `make deploy-*` command uses [helm](https://helm.sh/) and [skaffold](https://skaffold.dev/) to build and deploy `etcd-druid` to the target Kubernetes cluster. In addition to deploying `etcd-druid` it will also install the [Etcd CRD](https://github.com/gardener/etcd-druid/blob/3383e0219a6c21c6ef1d5610db964cc3524807c8/config/crd/bases/crd-druid.gardener.cloud_etcds.yaml) and [EtcdCopyBackup CRD](https://github.com/gardener/etcd-druid/blob/3383e0219a6c21c6ef1d5610db964cc3524807c8/config/crd/bases/crd-druid.gardener.cloud_etcdcopybackupstasks.yaml).

Reviewer: ```suggestion
Any variant of `make deploy-*` command uses [helm](https://helm.sh/) and [skaffold](https://skaffold.dev/) to build and deploy `etcd-druid` to the target Kubernetes cluster. In addition to deploying `etcd-druid` it will also install the [Etcd CRD](https://github.com/gardener/etcd-druid/blob/3383e0219a6c21c6ef1d5610db964cc3524807c8/config/crd/bases/crd-druid.gardener.cloud_etcds.yaml) and [EtcdCopyBackupsTask CRD](https://github.com/gardener/etcd-druid/blob/3383e0219a6c21c6ef1d5610db964cc3524807c8/config/crd/bases/crd-druid.gardener.cloud_etcdcopybackupstasks.yaml).
```

Author: Done

#### Debug mode

```bash
> make deploy-debug
```

This is similar to `make deploy-dev` but additionally configures containers in pods for debugging as required for each container's runtime technology. The associated debugging ports are exposed and labelled so that they can be port-forwarded to the local machine. Skaffold disables automatic image rebuilding and syncing when using the `debug` mode as compared to `dev` mode.

Go debugging uses [Delve](https://github.com/go-delve/delve). Please see the [skaffold debugging documentation](https://skaffold.dev/docs/workflows/debug/) how to setup your IDE accordingly. 

> **Note:** Resuming or stopping only a single goroutine (Go Issue [25578](https://github.com/golang/go/issues/25578), [31132](https://github.com/golang/go/issues/31132)) is currently not supported, so the action will cause all the goroutines to get activated or paused.

This means that when a goroutine is paused on a breakpoint, then all the other goroutines are also paused. This should be kept in mind when using `skaffold debug`.

## 03-Configure Backup [*Optional*]

> > **Note:** If you wish to do not backup any snapshots

Reviewer: This note seems incomplete. Did you want it to tell readers to not set `Etcd.Spec.Backup.Store` in order to disable snapshotting?

Author: removed it. I think this was a leftover

### Deploying a Local Backup Store Emulator

> **Note:** This section is ***Optional*** and is only meant to describe steps to deploy a local object store which can be used for testing and development. If you either do not wish to enable backups or you wish to use remote (infra-provider-specific) object store then this section can be skipped.

> An `Etcd` cluster provisioned via etcd-druid provides a capability to take regular delta and full snapshots and stored them in an object store. You can enable this functionality by ensuring that you fill in [spec.backup.store](https://github.com/gardener/etcd-druid/blob/3383e0219a6c21c6ef1d5610db964cc3524807c8/config/samples/druid_v1alpha1_etcd.yaml#L49-L54) section of the `Etcd` CR. 

Reviewer 2: ```suggestion
An `Etcd` cluster provisioned via etcd-druid provides a capability to take regular delta and full snapshots and store them in an object store. You can enable this functionality by ensuring that you fill in [spec.backup.store](https://github.com/gardener/etcd-druid/blob/3383e0219a6c21c6ef1d5610db964cc3524807c8/config/samples/druid_v1alpha1_etcd.yaml#L49-L54) section of the `Etcd` CR. 
```

| Backup Store Variant          | Setup Guide                                                |
| ----------------------------- | ---------------------------------------------------------- |
| Azure Object Storage Emulator | [Manage Azurite](manage-azurite-emulator.md) (Steps 00-03) |
| S3 Object Store Emulator      | [Manage LocalStack](manage-s3-emulator.md) (Steps 00-03)   |

### Setting up Cloud Provider Object Store Secret

> **Note:** This section is ***Optional***. If you have disabled backup functionality or if you are using local storage or one of the supported object store emulators then you can skip this section.

> A Kubernetes [Secret](https://kubernetes.io/docs/concepts/configuration/secret/) needs to be created for cloud provider Object Store access. You can refer to the Secret YAML templates [here](https://github.com/gardener/etcd-backup-restore/tree/master/example/storage-provider-secrets).  Replace the dummy values with the actual configuration and ensure that you have added the `metadata.name` and `metadata.namespace` to the secret.

Reviewer: We also need to add information about which permissions are to be configured for the acess credentials. For instance, the AWS S3 credential needs to have permissions like `s3:*` (or more restrictive permissions, once we evaluate those correctly for each provider). How would you like to go about this?

Author: Sure we can enhance the document once we have finalised the list of permissions that are required. We can then include the changes in the next documentation PR.

> **<u>Note</u>:**
>

> > * Secret should be deployed in the same namespace as `etcd-druid` deployment.

Reviewer 2: ```suggestion
> * Secret should be deployed in the same namespace as the Etcd resource.
```
It should rather be the Etcd CR namespace right?

Author: Thanks! corrected

If you wish to enable functionality to backup delta & full snapshots then uncomment `spec.backup.store` section.

```yaml
# Configuration for storage provider
store:
  secretRef:
    name: etcd-backup-secret-name
  container: object-storage-container-name
  provider: aws # options: aws,azure,gcp,openstack,alicloud,dell,openshift,local
  prefix: etcd-test
```

Brief explanation of the keys:

* `secretRef.name` is the name of the secret that was applied as mentioned above.
* `store.container` is the object storage bucket name.
* `store.provider` is the bucket provider. Pick from the options mentioned in comment.
* `store.prefix` is the folder name that you want to use for your snapshots inside the bucket.

> **Note:** For developer convenience we have provided object store emulator specific etcd CR variants which can be used as if as well.

## 05-Applying Etcd CR

Create the Etcd CR (Custom Resource) by applying the Etcd yaml to the cluster
```bash
> kubectl apply -f <path-to-etcd-cr-yaml>
```

## 06-Verify the Etcd Cluster

To obtain information on the etcd cluster you can invoke the following command:
```bash
> kubectl get etcd -o=wide
```

> We adhere to a naming convention for all resources that are provisioned for an `Etcd` cluster. Refer to [etcd-cluster-components](../../concepts/etcd-cluster-components.md) document to get details of all resources that re provisioned.

Reviewer: ```suggestion
We adhere to a naming convention for all resources that are provisioned for an `Etcd` cluster. Refer to [etcd-cluster-components](../../concepts/etcd-cluster-components.md) document to get details of all resources that are provisioned.
```

Author: Done

* If `--enable-etcd-spec-auto-reconcile` is switched off, then by default `Etcd` spec changes are not reconciled. One can intruct etcd-druid to reconcile an `Etcd` by putting `gardener.cloud/operation: reconcile` annotation on the `Etcd` CR. Once the reconciliation is successfully completed then this annotation will be removed by etcd-druid. You can use the following command to add this annotation:
  ```bash
  # Annotate etcd-test CR to reconcile
  kubectl annotate etcd etcd-test gardener.cloud/operation="reconcile"
  ```

## 08-Cleaning up the setup

If you wish to only delete the `Etcd` cluster then you can use the following command:
```bash
> kubectl delete etcd <etcd-name>
```

This will add the `deletionTimestamp` to the `Etcd` resource.  At the time the creation of the `Etcd` cluster, etcd-druid will add a finalizer to ensure that it cleans up all `Etcd` cluster resources before the CR is removed. 
```yaml
  finalizers:
  - druid.gardener.cloud/etcd-druid
```

etcd-druid will automatically pick up the deletion event and attempt clean up `Etcd` cluster resources. It will only remove the finaliser once all resources have been cleaned up.

> 

Reviewer: Please also add a point about `make undeploy` here to uninstall the etcd-druid installation on the cluster.

Author: sure. added

# Retrieve all key-value pairs from the etcd db
> etcdctl get --prefix ""
```

For a multi-node etcd cluster, insert the key-value pair using the `etcd` container of one etcd member and retrieve it from the `etcd` container of another member to verify consensus among the multiple etcd members.

## 07-Updating Etcd CR

`Etcd` CR can be updated with new changes. To ensure that `etcd-druid` reconciles the changes you can opt for one of the below options:

* While setting up druid if you have chosen to switch on `auto-reconcile` via CLI flag `--enable-etcd-spec-auto-reconcile` then any changes to the `Etcd` spec will be automatically reconciled.

* If `--enable-etcd-spec-auto-reconcile` is switched off, then by default `Etcd` spec changes are not reconciled. One can intruct etcd-druid to reconcile an `Etcd` by putting `gardener.cloud/operation: reconcile` annotation on the `Etcd` CR. Once the reconciliation is successfully completed then this annotation will be removed by etcd-druid. You can use the following command to add this annotation:
  ```bash
  # Annotate etcd-test CR to reconcile
  kubectl annotate etcd etcd-test gardener.cloud/operation="reconcile"

>   ```

Reviewer 2: For managing the etcd clusters, since you have already introduced a doc `managing-etcd-clusters` which explains this section in detail, can you just refer that?

Author: done
